org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/NavigateToTypeDefinitionHandler.java

```diff
@@ -0,0 +1,88 @@
+import java.util.List;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.IClassFile;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IMember;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.core.dom.IBinding;
+import org.eclipse.jdt.core.dom.ITypeBinding;
+import org.eclipse.jdt.core.dom.IVariableBinding;
+import org.eclipse.jdt.core.dom.NodeFinder;
+import org.eclipse.jdt.core.dom.SimpleName;
+import org.eclipse.jdt.core.manipulation.CoreASTProvider;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.lsp4j.Location;
+import org.eclipse.lsp4j.TextDocumentPositionParams;
+
+public class NavigateToTypeDefinitionHandler {
+
+	public NavigateToTypeDefinitionHandler() {
+	}
+
+	public List<? extends Location> typeDefinition(TextDocumentPositionParams position, IProgressMonitor monitor) {
+		ITypeRoot unit = JDTUtils.resolveTypeRoot(position.getTextDocument().getUri());
+		Location location = null;
+		if (unit != null && !monitor.isCanceled()) {
+			location = computeTypeDefinitionNavigation(unit, position.getPosition().getLine(), position.getPosition().getCharacter(), monitor);
+		}
+		return location == null ? null : Arrays.asList(location);
+	}
+
+	private Location computeTypeDefinitionNavigation(ITypeRoot unit, int line, int column, IProgressMonitor monitor) {
+		try {
+			CompilationUnit ast = CoreASTProvider.getInstance().getAST(unit, CoreASTProvider.WAIT_YES, null);
+			int offset = JsonRpcHelpers.toOffset(unit.getBuffer(), line, column);
+			if (ast == null || offset < 0) {
+				return null;
+			}
+			NodeFinder finder = new NodeFinder(ast, offset, 0);
+			ASTNode coveringNode = finder.getCoveringNode();
+			if (coveringNode instanceof SimpleName) {
+				IBinding resolvedBinding = ((SimpleName) coveringNode).resolveBinding();
+				if (resolvedBinding != null) {
+					ITypeBinding typeBinding = null;
+					if (resolvedBinding instanceof IVariableBinding) {
+						typeBinding = ((IVariableBinding) resolvedBinding).getType();
+					} else if (resolvedBinding instanceof ITypeBinding) {
+						typeBinding = (ITypeBinding) resolvedBinding;
+					}
+					if (typeBinding != null && typeBinding.getJavaElement() != null) {
+						IJavaElement element = typeBinding.getJavaElement();
+						ICompilationUnit compilationUnit = (ICompilationUnit) element.getAncestor(IJavaElement.COMPILATION_UNIT);
+						IClassFile cf = (IClassFile) element.getAncestor(IJavaElement.CLASS_FILE);
+						if (compilationUnit != null || (cf != null && cf.getSourceRange() != null)) {
+							return JDTUtils.toLocation(element);
+						}
+						if (element instanceof IMember && ((IMember) element).getClassFile() != null) {

```

Comment:
requires client to support classFileContents

Reply:
It is the similar situation for NavigateToDefinition. I have changed the test case for both. Both will run through the code to here.  But this line will handle the  case: 
```java
public static String toUri(IClassFile classFile) {
		if (JavaLanguageServerPlugin.getPreferencesManager() != null && !JavaLanguageServerPlugin.getPreferencesManager().isClientSupportsClassFileContent()) {
			return null;
		}
```